### ISSUE TITLE
Breaking 3.0 Changes

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#4]
There are actually quite a few breaking changes coming in 3.0, so I'm starting this issue to track them.

The one big thing that's not changing at all is how settings and default values are defined, so there should not be any changes required to your Settings classes.
### NFigAsyncStore merged into NFigStore

`NFigAsyncStore` has been removed. `NFigStore` now has both synchronous and asynchronous methods. All of the async methods are abstract, so must be implemented by inherited classes. The synchronous methods are virtual, and the default implementations use [Task.Run](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.run%28v=vs.110%29.aspx) to simply call the async methods and wait on the result. If you're implementing a store based on a database driver with native support for synchronous calls, you should consider providing your own implementation of the sync methods.
### NFigStore instances are Tier and DataCenter specific

You now have to provide the current tier and data center when initializing an NFigStore.

`GetAppSettings()` and `SubscribeToAppSettings()` no longer accept tier or data center parameters. Set and clear override methods no longer accept a tier argument, but do still accept a data center argument.

These changes were made for a few reasons:
1. To simplify the interfaces.
2. The idea of getting or setting overrides for a tier other than your current tier was dubious to begin with, unless you had multiple tiers connected to the same data source (a horrible idea). So having tier as an argument was misleading. [NFig.UI](https://github.com/NFig/NFig.UI) never supported editing other tiers.
3. With the addition of encrypted settings, it's expected that different tiers will use different encryption keys. Therefore, it wouldn't be possible to decrypt settings from a different tier, making it truly impossible to provide settings from another tier.
### User parameter

Set and clear override methods now have a `string user` parameter in order to support the new logging features. The parameter is not "optional", but you can pass `null` if you don't care about logging.

NFig is not opinionated about the format of the user string, and will accept any arbitrary string. Logging implementations should support strings of at least 100 [UCS-2](https://en.wikipedia.org/wiki/UTF-16) characters, as well as null.
### GetSettingInfo Removed

`.GetSettingInfo()`, and its async counterpart, have been removed. You can use `GetAllSettingInfos()` and then filter down to the individual SettingInfo that you need.
### GetAllSettingInfos only returns defaults relevant to the current tier

Previously, `GetAllSettingInfos()` returned all known default and override values. Because of the introduction of encrypted settings, it would be impossible to provide useful information about encrypted settings from another tier (unless all tiers shared the same encryption keys, which is a bad idea). The easiest way around this is to simply discard all defaults which are irrelevant to the current tier upon initialization.

> NFig will still _validate_ unencrypted defaults values for all tiers in order to catch as many errors as possible during local development, but it will not maintain any state about them after validation.
### Exceptions are no longer Generic

`InvalidSettingValueException` and `InvalidSettingOverridesException` no longer have type parameters. The `Tier` and `DataCenter` properties are also removed from `InvalidSettingValueException`. Tier is not applicable since it will always be the current tier of the application. DataCenter is still accessible in string form via `exception.Data["DataCenter"]`. The property was removed so that the exception types wouldn't need type arguments.
### SettingsFactory is internal

`SettingsFactory` is now an internal class. There is no reason to use it directly now that `NFigMemoryStore` is available.
### NFigStore Implementation Changes

There are numerous internal changes to what inheriting classes need to implement. Since the only implementation I'm aware of at this time is [NFig.Redis](https://github.com/NFig/NFig.Redis), I'm not going to worry about documenting them. If anyone has questions, they can ask.

**Comment by bretcope** [comment]
After the multi-tenancy refactor, the changes are simply too numerous to document. It will be easier to just describe how 3.0 works than to explain the differences between 2.0 and 3.0.
